### PR TITLE
Variant can now be copied into an already existing Variant

### LINKED ISSUE #752: Allow network variant copy to an existing variant
* **Do you want to request a *feature* or report a *bug*?**

Feature.

* **What is the current behavior?**

The only way of copying the content of a network variant is to create a new one with the `cloneVariant` method. As reported by @murgeyseb , it is in particular not possible to copy the content of a variant to the initial variant, since it necessarily preexists.

* **What is the expected behavior?**

We should be able to copy a variant to an existing one:
    - either with the existing `cloneVariant` method, maybe with an additional `mayOverwrite` parameter for security
    - or with a new `copyVariant` method

In any case, the copy code already exists in the `cloneVariant` method, so it is not much work.

* **What is the motivation / use case for changing the behavior?**

Being able to copy a network variant to the initial variant, or to any already existing variant. It may be useful to reintegrate the result of a process inside the initial variant.

* **Please tell us about your environment:**
PowSyBl Version: 2.4.1

### PR DESCRIPTION
**Pull request by AbdelHedhili**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
Fixes #752

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java [hunk 1]
```diff
@@ -56,6 +56,18 @@
...
+    /**
+     * Create or overwrite a variant by cloning an existing one.
+     *
+     * @param sourceVariantId the source variant id
+     * @param targetVariantIds the target variant id list (the ones that will be created)
+     * @throws com.powsybl.commons.PowsyblException
+     *                         if the source variant is not found or if a variant with
+     *                         an id of targetStateIds already exists
+     */
+    void cloneVariant(String sourceVariantId, List<String> targetVariantIds, boolean mayOverwrite);
```
Comment (MioRtia) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java]: add a default implementation in order not to break compatibility:
```java
default void cloneVariant(String sourceVariantId, List<String> targetVariantIds, boolean mayOverwrite) {
   cloneVariant(sourceVariantId, targetVariantIds);
}
```

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java [hunk 2]
```diff
@@ -67,6 +79,19 @@
...
+     *
+     * @param sourceVariantId the source variant id
+     * @param targetVariantId the target variant id list (the ones that will be created
+     * @param mayOverwrite indicate if the target can be overwritten if it is already existing
+     * @throws com.powsybl.commons.PowsyblException
+     *      *                         if the source variant is not found or if a variant with
+     *      *                         an id of targetStateIds already exists and the mayOverwrite
+     *      *                         parameter is set to false
+     */
+    void cloneVariant(String sourceVariantId, String targetVariantId, boolean mayOverwrite);
```
Comment (MioRtia) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java]: Same as above

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java [hunk 3]
```diff
@@ -56,6 +56,20 @@
...
+    /**
+     * Create or overwrite a variant by cloning an existing one.
+     *
+     * @param sourceVariantId the source variant id
+     * @param targetVariantIds the target variant id list (the ones that will be created)
+     * @throws com.powsybl.commons.PowsyblException
+     *                         if the source variant is not found or if a variant with
+     *                         an id of targetStateIds already exists
+     */
+    default void cloneVariant(String sourceVariantId, List<String> targetVariantIds, boolean mayOverwrite) {
```
Comment (mathbagu) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java]: Do we really want to keep the old method that throws an exception?

The javadoc is not OK: you forget to explain the third parameter. This method throws an exception only if the mayOverwrite is equal to false. Maybe the default implementation is not correct. Maybe we should add a try/catch to respect the contract of this method.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java [hunk 4]
```diff
@@ -67,6 +81,21 @@
...
+     * @param sourceVariantId the source variant id
+     * @param targetVariantId the target variant id list (the ones that will be created
+     * @param mayOverwrite indicate if the target can be overwritten if it is already existing
+     * @throws com.powsybl.commons.PowsyblException
+     *      *                         if the source variant is not found or if a variant with
+     *      *                         an id of targetStateIds already exists and the mayOverwrite
+     *      *                         parameter is set to false
+     */
+    default void cloneVariant(String sourceVariantId, String targetVariantId, boolean mayOverwrite) {
+        cloneVariant(sourceVariantId, targetVariantId);
```
Comment (mathbagu) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java]: I'm not sure this code is compliant with the javadoc: you should probably add a try/catch here two.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java (hunk 3 above)
Comment (sylvlecl) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java]: I think we should either have no default implementation or throw an `UnsupportedOperationException` :
in any case the default implementation will not respect the contract (if you catch the exception, the variant is not copied anyway).

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java [hunk 5]
```diff
@@ -128,6 +142,18 @@ public void cloneVariant(String sourceVariantId, List<String> targetVariantIds)
...
+
+        if (extendedCount > 0) {
+            for (MultiVariantObject obj : getStafulObjects()) {
+                obj.extendVariantArraySize(initVariantArraySize, extendedCount, sourceIndex);
+            }
+            LOGGER.trace("Extending variant array size to {} (+{})", variantArraySize, extendedCount);
+        }
+    }
+
+    private void allocateVariantArrayElements(Integer sourceIndex, List<Integer> recycled, List<Integer> overwritten) {
```
Comment (ChamseddineBhd) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java]: same remark

**Comment by ChamseddineBhd** [comment]
Should we make the old clone methods deprecated or not ?
@AbdelHedhili  @mathbagu ? 
For me i think we should do that, our user should use the clone method with the new signature so he'll be aware of that problem;

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java [hunk 6]
```diff
@@ -112,11 +122,15 @@ public void cloneVariant(String sourceVariantId, List<String> targetVariantIds)
         int initVariantArraySize = variantArraySize;
         int extendedCount = 0;
         List<Integer> recycled = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (AbdelHedhili) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java]: I don't agree on that. The current names describe exactly what their goals are, you know that one array contains the variants that will be "overwritten", in the other it'll be the ones that will get "recycled". So for my part, I don't agree with this modification. I think we need another point of view to help us decide whether or not we are changing them. If we really want to give more info we could use : recycledVariantIndexes and overwrittenVariantIndexes.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java [hunk 7]
```diff
@@ -56,6 +56,21 @@
      */
     void cloneVariant(String sourceVariantId, List<String> targetVariantIds);
 
+    /**
+     * Create or overwrite a variant by cloning an existing one.
+     *
+     * @param sourceVariantId the source variant id
+     * @param targetVariantIds the target variant id list (the ones that will be created)
+     * @param mayOverwrite indicate if the target can be overwritten if it is already existing
```
[the remark this answers is not in the document]
Comment (MioRtia) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java]: I don't agree with this remark, I think the original wording is more comprehensive.

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ChamseddineBhd) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java]: i have discussed that avec miora, i'm convinced now with that name :)

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ChamseddineBhd) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java]: @MioRtia  at least :  indicate ->  indicates

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java [hunk 8]
```diff
@@ -137,11 +163,14 @@ public void cloneVariant(String sourceVariantId, List<String> targetVariantIds)
                 LOGGER.trace("Recycling variant array indexes {}", Arrays.toString(indexes));
             }
         }
-        if (extendedCount > 0) {
+        if (!overwritten.isEmpty()) {
```
Comment (sylvlecl) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java]: Ok but we could probably have merged the 2 blocks into only one, they are identical except for the log.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java [hunk 9]
```diff
@@ -56,6 +56,19 @@
...
+     * @param mayOverwrite indicates if the target can be overwritten if it is already existing
+     * @throws com.powsybl.commons.PowsyblException
+     *                       if a variant with an id of targetVariantIds already exists and
+     *                       the mayOverwrite parameter is set to false
+     */
+    void cloneVariant(String sourceVariantId, List<String> targetVariantIds, boolean mayOverwrite);
+
+
     /**
      * Create a new variant by cloning an existing one.
```
[the remark this answers is not in the document]
Comment (AbdelHedhili) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java]: I don't understand : it's already said in the javadoc 
"@throws com.powsybl.commons.PowsyblException
     *                       if a variant with an id of targetVariantIds already exists and
     *                       the mayOverwrite parameter is set to {@code false}"

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java (hunk 8 above)
Comment (AbdelHedhili) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java]: I don't see how to merge the two blocks without making it less understandable

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (sylvlecl) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/VariantManager.java]: You're right, I got confused by the diff !

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java (hunk 8 above)
Comment (sylvlecl) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VariantManagerImpl.java]: Ok!

**Pull request merged**